### ISSUE TITLE
Show, unbind, and delete unbound routes

### ISSUE THREAD

**Issue by suprenant** [18F/cg-dashboard#582]
As a developer or project manager, I can see, unbind, and delete unbound routes.

Notes: 
 - Currently showing number of bound routes
 - If showing 3/1000, only 1 will be shown in the UI
 - Want to be clear out of how many remain
 - Other apps may not be consuming the routes but could be impacting the quota
 - Can be bound and unbound

May need some design work

Acceptance criteria: 
 - [ ] Routes panel should show remaining routes available on space
 - [ ] Routes panel should show if such a limit/quota on routes exists
 - [ ] There should be an action to unbind a route
 - [ ] There should be an action to delete a route that is unbound

**Comment by suprenant** [comment]
Next step: have a spike so @msecret can discuss w/team

**Comment by thisisdano** [comment]
I suggest something like `N routes exist in your [space/org quota OR application limit]` if such a limit exists; otherwise show no notice

**Comment by msecret** [comment]
Another issue I thought of here that isn't just related to the quota number.

Let's say marco.18f.gov route exists on my staging space.

On my app page I have liberator.18f.gov as a route. I want to add the marco.18f.gov route to the app, so I click the add button and fill in marco.18f.gov. There would be an error condition that the route already exists because it already exists on my staging space.

**Comment by suprenant** [comment]
@thisisdano @vz can you take a look at the @msecret comment above and provide design so we can move this issue forward?

**Comment by suprenant** [comment]
@thisisdano @vz3 bumping again here. Slacked you the same. Can we work to resolve our design questions so we can pick up this issue?

**Comment by vz3** [comment]
@msecret How would an unbound route be created? Can you tell me why we wouldn't just delete routes that are unbound (i.e. routes can only exist when bound. If they are unbound they are just deleted).

**Comment by vz3** [comment]
@msecret If unbound routes must exist I think we need to show all unbound routes for a particular org/space(?) when users are adding a new route. Or if users add an unbound route, can we just bind it to the requested app/space rather than generate that error?

**Comment by msecret** [comment]
@vz3 an unbound route could be created with the CLI. I've also seen cloud foundry UIs that allow creation of a route at the space level, which would be unbound.

**Comment by msecret** [comment]
I think we wouldn't delete unbound routes because they could be bound to more then one app. But if a route isn't used in any apps, then there isn't much reason to keep it around.

**Comment by msecret** [comment]
Another thing: routes are created at the space level, not app level. So our options are to show unbound routes for the space on the app page or we could just change the functionality to only bind the route when they try and create a new one.

I like the 2nd idea as we're modifying the UI in the way we think will be most successful with people. If we go with that route, what about the route quota which we discussed above. Should we just changed the wording to something more like `N routes exist in your space`? Or something similar to explain that number is related to all the routes in your space and not just the bound routes you see here?

**Comment by suprenant** [comment]
FYI team is planning a sketch session on this tomorrow

@vz3

**Comment by thisisdano** [comment]
Couple prelim sketches for this.

![image](https://cloud.githubusercontent.com/assets/11464021/18529763/53235b52-7a83-11e6-80af-a9aca1ef0741.png)

**Comment by suprenant** [comment]
For reference, doc from our session working through this issue: https://docs.google.com/document/d/1I9w_9cDjbO5d-W21FDIzfSG1nN2q3ps9rmHXRCJuX-s/edit

**Comment by suprenant** [comment]
@thisisdano as discussed, let's target Monday for your design revisions here.

**Comment by thisisdano** [comment]
I spent the weekend trying to think through a clean design direction for the dashboard. I don't think I'm there yet, but here's where I am now.

I'll flesh this out throughout the day

![image](https://cloud.githubusercontent.com/assets/11464021/18640624/c359f8a4-7e4d-11e6-88f6-5860a8ee8b1c.png)

**Comment by vz3** [comment]
Yay for visual consistency here!

**Comment by vz3** [comment]
Tracking some questions:
- How verbose should we be in button language? Any thoughts on this @brittag?
- When do we use 'x' to delete, when do we use trash icon?
- Can services be bound to multiple apps?
- How do we present system errors here (i.e. The service has failed to bind)
- Does the marketplace link for services work?

**Comment by thisisdano** [comment]
Prototype for this functionality in progress here: https://gsa.invisionapp.com/d/main#/console/8652272/189837631/preview

**Comment by mogul** [comment]
@vz3: 

> - Can services be bound to multiple apps?

Yes, absolutely.

**Comment by thisisdano** [comment]
@mogul 
Oh, I could have sworn we heard otherwise from @msecret !
(That was the way I'd set it up first.)
I must have misunderstood...

**Comment by mogul** [comment]
Here's a simple example: One app in the space records user requests into a bound database service. A second app in the same space bound to the same database service asynchonously reads the requests out, handles them and mails the users when the request is completed.

**Comment by thisisdano** [comment]
That certainly makes sense. I think we can signal that info in the app panel and (potentially) give more contextual insight on the space service management page. (Which, of course, does not exist yet.)

**Comment by thisisdano** [comment]
Updated Invision to show if an unbound service is bound to another app (which would prevent delete action in the app services panel).
https://gsa.invisionapp.com/share/9T8GMJ8B8#/190897181_App-Services-6c--default

<img width="496" alt="in-use" src="https://cloud.githubusercontent.com/assets/11464021/18759606/146325ac-80b2-11e6-95ac-3dcd60610e20.png">

**Comment by mogul** [comment]
I like it! However some services will be free, so "Purchase" as a generic action is not appropriate. I'd go with something like "Provision" or "Create" or "Add", where the free/paid status is on the next form they see.

**Comment by thisisdano** [comment]
👍

**Comment by msecret** [comment]
Currently I'm only showing services and routes that are completely unbound on the app page as available, so at least nothing would be breaking. Can update to this though.

**Comment by mogul** [comment]
It probably goes without saying, but: This is also a cue as to the kind of info that could appear on a space-details view: all the services in the space, all the routes in the space, which ones are bound and to how many/which apps, the ability to add a route or space, etc. 

I do think that most of the time action involving services and routes happens from an app-centric context, so it's good that we've been concentrating there. In terms of conceptual hierarchy, though, the CRUD of services and routes would be exposed at the same level as apps within a space. Note the availability of `cf create-route` (space-level) and `cf map-route` (app-level), for example, even though it's common to set up a route in tandem with pushing a new app (`cf push` and `manifest.yml` include the parameters for this).

**Comment by thisisdano** [comment]
YES. Agreed

**Comment by thisisdano** [comment]
The little 'Manage Routes in [space name]' link at the top of the panel
would go to the space level management you outline

**Comment by mogul** [comment]
Another thing to consider: There are some services which are not bound _to the app_, but rather _to the traffic flowing to the app, embodied by the route_... These are the so-called "route services", which basically act as proxies. For example, in the GovCloud environment we now have these two:

- uaa-auth - Secure a route with cloud.gov authentication
  - This is what makes users authenticate via GSA SecureAuth when visiting Tock
- cdn-route - Pipe traffic to a given route through a CDN
  - This is what makes it easy to [set up DNS in GovCloud](https://docs.cloud.gov/apps/custom-domains/#managed-service-method)

There are more route services planned in the backlog.

There are separate commands for binding route-services to routes: `bind-route-service` and `unbind-route-service`. The CRUD for these services basically matches that of any other service (with plans, parameters, etc), but they get bound to routes rather than apps. So it's probably worth thinking about how to expose that interaction, likely as a separate story if you haven't already got one.

**Comment by thisisdano** [comment]
^^^ Now there's something to think about! I gotta wrap my head around this for a bit to consider how we'd address it.
@msecret @vz3 @suprenant

**Comment by mogul** [comment]
Note that route-service bindings don't involve apps at all, they're just about establishing/dissolving the relationship between the service and the route. What app the route happens to be pointing to is immaterial to that interaction. The reason I'm bringing it up is that it's almost certainly an interaction that would happen in the context of the space, where we don't do much but list apps now.

**Comment by msecret** [comment]
Waiting on https://github.com/18F/cg-style/pull/155, https://github.com/18F/cg-style/pull/154

**Comment by thisisdano** [comment]
@suprenant The issues that @mogul is bringing up here should be flagged for a new story/ies — specifically for route and service management at the space level.

**Comment by suprenant** [comment]
@thisisdano sounds good. As discussed on Slack, we'll talk and groom around this a bit tomorrow. Thanks for highlighting it!

**Comment by suprenant** [comment]
Update from earlier today: @msecret working on Services bound to more than one app
